# **Notebook Purpose**
*In this notebook we will using all datasets (tour-guide and tourist) as tour-guide. Because we want get more possibility match guide for users(tourist)*

In [1]:
import sys
import pandas as pd
import numpy as np
import tensorflow as tf

from sklearn.metrics.pairwise import euclidean_distances

# **Tidying Up Datasets**

## Load Datasets

In [2]:
# Load Tour Guide Datasets
guide_df = pd.read_csv('data/tour-guide-data.csv')
guide_df

Name  Gender  Domicile  17-25  26-34  35-43  44-52   
0              Nabhan Nabilah       1         0      1      0      0      0  \
1                  Salsabilla       1         0      1      0      0      0   
2                     Maulani       1         0      1      0      0      0   
3               Abiyyu Farhan       0         0      1      0      0      0   
4            Rahmaliyah Kadir       1         1      1      0      0      0   
5                      Faizah       1         1      1      0      0      0   
6                      Andani       1         1      0      1      0      0   
7               Muthiah Hanun       1         1      0      1      0      0   
8         Ihlasul Mufti Faqih       0         1      1      0      0      0   
9   Dessylva Maharany Santosa       1         0      1      0      0      0   
10                     Elma N       1         0      1      0      0      0   
11                       Hani       1         1      1      0      0      0   
12             Jasmine Khaira       1         0      1      0      0      0   
13                    fadilah       1         1      1      0      0      0   
14                      Winda       1         0      1      0      0      0   
15         Syifa Choerunnisa        1         0      1      0      0      0   
16                       Pebi       1         0      1      0      0      0   
17                  Nur Aliya       1         0      1      0      0      0   
18               zahra kamila       1         0      1      0      0      0   
19                     Hafidz       0         0      1      0      0      0   
20                        Dwi       1         0      1      0      0      0   
21                          E       1         0      1      0      0      0   
22                     debbie       1         0      0      1      0      0   
23                       Dera       1         0      1      0      0      0   
24           Salsabilla Rizki       1         0      1      0      0      0   
25                      Salma       1         0      1      0      0      0   
26                       Deva       1         0      0      1      0      0   
27                      Dzaky       0         0      1      0      0      0   

    52+  Historical tours  Adventure tours  ...  CSN1  CSN2  CSN3  CSN4  CSN5   
0     0                 0                0  ...     5     2     5     5     5  \
1     0                 0                1  ...     4     3     3     4     4   
2     0                 1                0  ...     4     4     3     4     4   
3     0                 0                1  ...     4     5     3     5     4   
4     0                 0                0  ...     4     3     3     4     3   
5     0                 0                0  ...     4     2     5     5     5   
6     0                 0                1  ...     4     4     4     3     3   
7     0                 0                0  ...     5     5     5     5     5   
8     0                 0                1  ...     5     4     5     4     3   
9     0                 1                0  ...     3     3     3     4     3   
10    0                 0                1  ...     3     3     4     3     3   
11    0                 0                0  ...     4     4     4     5     5   
12    0                 1                1  ...     5     3     3     3     2   
13    0                 0                1  ...     4     3     3     3     2   
14    0                 1                0  ...     4     3     3     4     4   
15    0                 1                0  ...     3     5     5     5     5   
16    0                 0                0  ...     4     4     4     4     4   
17    0                 0                0  ...     4     4     4     5     5   
18    0                 1                0  ...     4     3     3     4     3   
19    0                 1                1  ...     2     1     1     1     2   
20    0                 0       

In [3]:
# Add roles column as a guide
guide_df['Roles'] = 'guide'
guide_df

Name  Gender  Domicile  17-25  26-34  35-43  44-52   
0              Nabhan Nabilah       1         0      1      0      0      0  \
1                  Salsabilla       1         0      1      0      0      0   
2                     Maulani       1         0      1      0      0      0   
3               Abiyyu Farhan       0         0      1      0      0      0   
4            Rahmaliyah Kadir       1         1      1      0      0      0   
5                      Faizah       1         1      1      0      0      0   
6                      Andani       1         1      0      1      0      0   
7               Muthiah Hanun       1         1      0      1      0      0   
8         Ihlasul Mufti Faqih       0         1      1      0      0      0   
9   Dessylva Maharany Santosa       1         0      1      0      0      0   
10                     Elma N       1         0      1      0      0      0   
11                       Hani       1         1      1      0      0      0   
12             Jasmine Khaira       1         0      1      0      0      0   
13                    fadilah       1         1      1      0      0      0   
14                      Winda       1         0      1      0      0      0   
15         Syifa Choerunnisa        1         0      1      0      0      0   
16                       Pebi       1         0      1      0      0      0   
17                  Nur Aliya       1         0      1      0      0      0   
18               zahra kamila       1         0      1      0      0      0   
19                     Hafidz       0         0      1      0      0      0   
20                        Dwi       1         0      1      0      0      0   
21                          E       1         0      1      0      0      0   
22                     debbie       1         0      0      1      0      0   
23                       Dera       1         0      1      0      0      0   
24           Salsabilla Rizki       1         0      1      0      0      0   
25                      Salma       1         0      1      0      0      0   
26                       Deva       1         0      0      1      0      0   
27                      Dzaky       0         0      1      0      0      0   

    52+  Historical tours  Adventure tours  ...  CSN2  CSN3  CSN4  CSN5  OPN1   
0     0                 0                0  ...     2     5     5     5     2  \
1     0                 0                1  ...     3     3     4     4     2   
2     0                 1                0  ...     4     3     4     4     3   
3     0                 0                1  ...     5     3     5     4     4   
4     0                 0                0  ...     3     3     4     3     3   
5     0                 0                0  ...     2     5     5     5     3   
6     0                 0                1  ...     4     4     3     3     4   
7     0                 0                0  ...     5     5     5     5     5   
8     0                 0                1  ...     4     5     4     3     2   
9     0                 1                0  ...     3     3     4     3     2   
10    0                 0                1  ...     3     4     3     3     2   
11    0                 0                0  ...     4     4     5     5     4   
12    0                 1                1  ...     3     3     3     2     3   
13    0                 0                1  ...     3     3     3     2     2   
14    0                 1                0  ...     3     3     4     4     4   
15    0                 1                0  ...     5     5     5     5     5   
16    0                 0                0  ...     4     4     4     4     4   
17    0                 0                0  ...     4     4     5     5     3   
18    0                 1                0  ...     3     3     4     3     3   
19    0                 1                1  ...     1     1     1     2     2   
20    0                 0       

In [4]:
# Load Tourist Datasets
tourist_df = pd.read_csv('data/tourist-data.csv')
tourist_df

Names  Genders  Destinations  17-25  26-34  35-43   
0    Novia Rizki Wulandari         1             0      1      0      0  \
1   Herlina Kusyanuri Putri        1             0      1      0      0   
2              Rifdah Alyaa        1             0      1      0      0   
3            Alexis Purnomo        0             0      0      0      0   
4      Rifky Surya Pratama         0             0      1      0      0   
..                      ...      ...           ...    ...    ...    ...   
68                      Féi        1             1      1      0      0   
69                   Annisa        1             1      1      0      0   
70                   Usamah        0             1      1      0      0   
71                   Helena        1             0      0      0      0   
72                     Biru        1             0      1      0      0   

    44-52  52+  Historical tours  Adventure tours  ...  CSN1  CSN2  CSN3   
0       0    0                 0                1  ...     5     4     4  \
1       0    0                 1                0  ...     5     5     5   
2       0    0                 0                0  ...     2     3     2   
3       0    0                 0                1  ...     3     5     3   
4       0    0                 0                1  ...     4     3     3   
..    ...  ...               ...              ...  ...   ...   ...   ...   
68      0    0                 0                0  ...     4     2     4   
69      0    0                 0                0  ...     4     2     3   
70      0    0                 0                1  ...     5     5     4   
71      0    0                 1                0  ...     4     3     4   
72      0    0                 0                0  ...     4     3     3   

    CSN4  CSN5  OPN1  OPN2  OPN3  OPN4  OPN5  
0      5     5     4     4     2     4     4  
1      5     5     4     3     3     4     3  
2      3     4     3     2     3     4     3  
3      4     5     4     4     1     4     4  
4      4     4     2     3     3     2     3  
..   ...   ...   ...   ...   ...   ...   ...  
68     4     4     3     4     2     4     4  
69     3     3     3     3     2     4     3  
70     5     5     4     4     5     5     5  
71     4     4     3     4     2     5     4  
72     4     2     4     3     3     3     3  

[73 rows x 41 columns]

In [5]:
# Add roles columns as a tourist
tourist_df['Roles'] = 'guide'
tourist_df

Names  Genders  Destinations  17-25  26-34  35-43   
0    Novia Rizki Wulandari         1             0      1      0      0  \
1   Herlina Kusyanuri Putri        1             0      1      0      0   
2              Rifdah Alyaa        1             0      1      0      0   
3            Alexis Purnomo        0             0      0      0      0   
4      Rifky Surya Pratama         0             0      1      0      0   
..                      ...      ...           ...    ...    ...    ...   
68                      Féi        1             1      1      0      0   
69                   Annisa        1             1      1      0      0   
70                   Usamah        0             1      1      0      0   
71                   Helena        1             0      0      0      0   
72                     Biru        1             0      1      0      0   

    44-52  52+  Historical tours  Adventure tours  ...  CSN2  CSN3  CSN4   
0       0    0                 0                1  ...     4     4     5  \
1       0    0                 1                0  ...     5     5     5   
2       0    0                 0                0  ...     3     2     3   
3       0    0                 0                1  ...     5     3     4   
4       0    0                 0                1  ...     3     3     4   
..    ...  ...               ...              ...  ...   ...   ...   ...   
68      0    0                 0                0  ...     2     4     4   
69      0    0                 0                0  ...     2     3     3   
70      0    0                 0                1  ...     5     4     5   
71      0    0                 1                0  ...     3     4     4   
72      0    0                 0                0  ...     3     3     4   

    CSN5  OPN1  OPN2  OPN3  OPN4  OPN5  Roles  
0      5     4     4     2     4     4  guide  
1      5     4     3     3     4     3  guide  
2      4     3     2     3     4     3  guide  
3      5     4     4     1     4     4  guide  
4      4     2     3     3     2     3  guide  
..   ...   ...   ...   ...   ...   ...    ...  
68     4     3     4     2     4     4  guide  
69     3     3     3     2     4     3  guide  
70     5     4     4     5     5     5  guide  
71     4     3     4     2     5     4  guide  
72     2     4     3     3     3     3  guide  

[73 rows x 42 columns]

## Combine Datasets Into Users Datasets

In [6]:
# Check if column names are the same
columns_equal = guide_df.columns.equals(tourist_df.columns)

if columns_equal:
    print("The DataFrames have the same column names.")
else:
    print("The DataFrames do not have the same column names.")

The DataFrames do not have the same column names.


In [7]:
# Get column names from guide_df and tourist_df
guide_columns = set(guide_df.columns)
tourist_columns = set(tourist_df.columns)

# Find the different columns
diff_col_in_guide = guide_columns - tourist_columns
diff_col_in_tourist = tourist_columns - guide_columns

# Print the different columns
print("Different columns in guide_df:", diff_col_in_guide)
print("Different columns in tourist_df:", diff_col_in_tourist)


Different columns in guide_df: {'Domicile', 'Name', 'Gender'}
Different columns in tourist_df: {'Names', 'Destinations', 'Genders'}


In [8]:
# Rename the diff column
guide_df = guide_df.rename(columns={'Name': 'Names'})
guide_df = guide_df.rename(columns={'Gender': 'Genders'})
guide_df = guide_df.rename(columns={'Domicile': 'Destinations'})


In [9]:
# Check if column names are the same
columns_equal = guide_df.columns.equals(tourist_df.columns)

if columns_equal:
    print("The DataFrames have the same column names.")
else:
    print("The DataFrames do not have the same column names.")

The DataFrames have the same column names.


In [10]:
user_df = pd.concat([tourist_df, guide_df])
user_df = user_df.reset_index(drop=True)
user_df

Names  Genders  Destinations  17-25  26-34  35-43   
0     Novia Rizki Wulandari         1             0      1      0      0  \
1    Herlina Kusyanuri Putri        1             0      1      0      0   
2               Rifdah Alyaa        1             0      1      0      0   
3             Alexis Purnomo        0             0      0      0      0   
4       Rifky Surya Pratama         0             0      1      0      0   
..                       ...      ...           ...    ...    ...    ...   
96                      Dera        1             0      1      0      0   
97          Salsabilla Rizki        1             0      1      0      0   
98                     Salma        1             0      1      0      0   
99                      Deva        1             0      0      1      0   
100                    Dzaky        0             0      1      0      0   

     44-52  52+  Historical tours  Adventure tours  ...  CSN2  CSN3  CSN4   
0        0    0                 0                1  ...     4     4     5  \
1        0    0                 1                0  ...     5     5     5   
2        0    0                 0                0  ...     3     2     3   
3        0    0                 0                1  ...     5     3     4   
4        0    0                 0                1  ...     3     3     4   
..     ...  ...               ...              ...  ...   ...   ...   ...   
96       0    0                 0                0  ...     4     5     5   
97       0    0                 0                1  ...     5     5     5   
98       0    0                 1                1  ...     5     5     5   
99       0    0                 0                1  ...     4     4     3   
100      0    0                 0                1  ...     5     5     5   

     CSN5  OPN1  OPN2  OPN3  OPN4  OPN5  Roles  
0       5     4     4     2     4     4  guide  
1       5     4     3     3     4     3  guide  
2       4     3     2     3     4     3  guide  
3       5     4     4     1     4     4  guide  
4       4     2     3     3     2     3  guide  
..    ...   ...   ...   ...   ...   ...    ...  
96      5     4     5     5     5     5  guide  
97      5     5     5     2     3     4  guide  
98      5     5     5     4     5     4  guide  
99      4     3     4     2     4     3  guide  
100     5     5     5     5     4     5  guide  

[101 rows x 42 columns]

## Check Summary New Datasets

In [11]:
# Looking at the index of the columns
columns = user_df.columns
for idx, column in enumerate(columns):
    print(f"Column index: {idx}, Column name: {column}")

Column index: 0, Column name: Names
Column index: 1, Column name: Genders
Column index: 2, Column name: Destinations
Column index: 3, Column name: 17-25
Column index: 4, Column name: 26-34
Column index: 5, Column name: 35-43
Column index: 6, Column name: 44-52
Column index: 7, Column name: 52+
Column index: 8, Column name: Historical tours
Column index: 9, Column name: Adventure tours
Column index: 10, Column name: Nature and wildlife tours
Column index: 11, Column name: Culinary tours
Column index: 12, Column name: Wellness and retreat tours
Column index: 13, Column name: Architectural tours
Column index: 14, Column name: Educational tours
Column index: 15, Column name: Shopping tours
Column index: 16, Column name: EXT1
Column index: 17, Column name: EXT2
Column index: 18, Column name: EXT3
Column index: 19, Column name: EXT4
Column index: 20, Column name: EXT5
Column index: 21, Column name: EST1
Column index: 22, Column name: EST2
Column index: 23, Column name: EST3
Column index: 24,

In [12]:
user_df.describe()

Genders  Destinations       17-25       26-34  35-43  44-52    52+   
count  101.000000    101.000000  101.000000  101.000000  101.0  101.0  101.0  \
mean     0.702970      0.415842    0.702970    0.049505    0.0    0.0    0.0   
std      0.459229      0.495325    0.459229    0.218002    0.0    0.0    0.0   
min      0.000000      0.000000    0.000000    0.000000    0.0    0.0    0.0   
25%      0.000000      0.000000    0.000000    0.000000    0.0    0.0    0.0   
50%      1.000000      0.000000    1.000000    0.000000    0.0    0.0    0.0   
75%      1.000000      1.000000    1.000000    0.000000    0.0    0.0    0.0   
max      1.000000      1.000000    1.000000    1.000000    0.0    0.0    0.0   

       Historical tours  Adventure tours  Nature and wildlife tours  ...   
count        101.000000       101.000000                 101.000000  ...  \
mean           0.386139         0.475248                   0.653465  ...   
std            0.489291         0.501878                   0.478239  ...   
min            0.000000         0.000000                   0.000000  ...   
25%            0.000000         0.000000                   0.000000  ...   
50%            0.000000         0.000000                   1.000000  ...   
75%            1.000000         1.000000                   1.000000  ...   
max            1.000000         1.000000                   1.000000  ...   

             CSN1        CSN2        CSN3        CSN4        CSN5        OPN1   
count  101.000000  101.000000  101.000000  101.000000  101.000000  101.000000  \
mean     3.683168    3.396040    3.613861    3.811881    3.841584    3.297030   
std      1.009264    1.010734    1.009656    1.046068    0.945861    0.964827   
min      1.000000    1.000000    1.000000    1.000000    1.000000    2.000000   
25%      3.000000    3.000000    3.000000    3.000000    3.000000    3.000000   
50%      4.000000    3.000000    4.000000    4.000000    4.000000    3.000000   
75%      4.000000    4.000000    4.000000    5.000000    5.000000    4.000000   
max      5.000000    5.000000    5.000000    5.000000    5.000000    5.000000   

             OPN2        OPN3        OPN4        OPN5  
count  101.000000  101.000000  101.000000  101.000000  
mean     3.435644    2.910891    3.841584    3.445545  
std      0.973816    1.040183    0.977064    0.994739  
min      1.000000    1.000000    1.000000    1.000000  
25%      3.000000    2.000000    3.000000    3.000000  
50%      3.000000    3.000000    4.000000    3.000000  
75%      4.000000    3.000000    5.000000    4.000000  
max      5.000000    5.000000    5.000000    5.000000  

[8 rows x 40 columns]

In [13]:
# Check that we not missing any value after combine columns
print('Is there any missing value? ', user_df.isnull().values.any())
print('How many missing values? ', user_df.isnull().values.sum())

Is there any missing value?  False
How many missing values?  0


# **Concatenating Personality Type Prediction**

In [14]:
# Copy origin user data into match data to use in cluster
match_df = user_df.copy()

In [15]:
# Drop Column that not use for clustering match
match_df.drop(['Names', 'Destinations', 'Roles'], axis=1, inplace=True)
match_df

Genders  17-25  26-34  35-43  44-52  52+  Historical tours   
0          1      1      0      0      0    0                 0  \
1          1      1      0      0      0    0                 1   
2          1      1      0      0      0    0                 0   
3          0      0      0      0      0    0                 0   
4          0      1      0      0      0    0                 0   
..       ...    ...    ...    ...    ...  ...               ...   
96         1      1      0      0      0    0                 0   
97         1      1      0      0      0    0                 0   
98         1      1      0      0      0    0                 1   
99         1      0      1      0      0    0                 0   
100        0      1      0      0      0    0                 0   

     Adventure tours  Nature and wildlife tours  Culinary tours  ...  CSN1   
0                  1                          1               0  ...     5  \
1                  0                          0               1  ...     5   
2                  0                          1               1  ...     2   
3                  1                          1               0  ...     3   
4                  1                          1               1  ...     4   
..               ...                        ...             ...  ...   ...   
96                 0                          1               1  ...     5   
97                 1                          1               1  ...     3   
98                 1                          0               1  ...     5   
99                 1                          1               1  ...     4   
100                1                          0               0  ...     5   

     CSN2  CSN3  CSN4  CSN5  OPN1  OPN2  OPN3  OPN4  OPN5  
0       4     4     5     5     4     4     2     4     4  
1       5     5     5     5     4     3     3     4     3  
2       3     2     3     4     3     2     3     4     3  
3       5     3     4     5     4     4     1     4     4  
4       3     3     4     4     2     3     3     2     3  
..    ...   ...   ...   ...   ...   ...   ...   ...   ...  
96      4     5     5     5     4     5     5     5     5  
97      5     5     5     5     5     5     2     3     4  
98      5     5     5     5     5     5     4     5     4  
99      4     4     3     4     3     4     2     4     3  
100     5     5     5     5     5     5     5     4     5  

[101 rows x 39 columns]

In [16]:
persona_test_result = match_df.iloc[:, -25:]
persona_test_result

EXT1  EXT2  EXT3  EXT4  EXT5  EST1  EST2  EST3  EST4  EST5  ...  CSN1   
0       4     2     4     4     3     3     3     1     2     3  ...     5  \
1       3     3     1     1     2     2     4     4     5     3  ...     5   
2       3     4     2     1     2     3     2     2     4     3  ...     2   
3       3     4     5     3     2     1     1     1     1     1  ...     3   
4       4     4     4     5     2     3     4     4     3     2  ...     4   
..    ...   ...   ...   ...   ...   ...   ...   ...   ...   ...  ...   ...   
96      5     4     5     4     5     4     3     3     5     4  ...     5   
97      3     4     3     3     3     2     4     2     2     2  ...     3   
98      2     1     4     4     4     4     5     5     2     1  ...     5   
99      4     4     4     4     4     2     3     2     4     3  ...     4   
100     4     5     5     5     5     1     1     1     1     4  ...     5   

     CSN2  CSN3  CSN4  CSN5  OPN1  OPN2  OPN3  OPN4  OPN5  
0       4     4     5     5     4     4     2     4     4  
1       5     5     5     5     4     3     3     4     3  
2       3     2     3     4     3     2     3     4     3  
3       5     3     4     5     4     4     1     4     4  
4       3     3     4     4     2     3     3     2     3  
..    ...   ...   ...   ...   ...   ...   ...   ...   ...  
96      4     5     5     5     4     5     5     5     5  
97      5     5     5     5     5     5     2     3     4  
98      5     5     5     5     5     5     4     5     4  
99      4     4     3     4     3     4     2     4     3  
100     5     5     5     5     5     5     5     4     5  

[101 rows x 25 columns]

In [17]:
# Convert input data to float32
data_input = np.array(persona_test_result, dtype=np.float32)

# Load model
model_path = 'persona_model/1'
loaded_model = tf.saved_model.load(model_path)

# Get the default signature of the model
inference_fn = loaded_model.signatures["serving_default"]

# Use models to make persona type predictions
predictions = inference_fn(tf.constant(data_input))["dense_2"]

In [18]:
predicted_clusters = np.argmax(predictions, axis=1)
persona_type = pd.DataFrame({'Persona type': predicted_clusters})
persona_type 

Persona type
0               2
1               1
2               3
3               2
4               4
..            ...
96              2
97              2
98              0
99              0
100             2

[101 rows x 1 columns]

In [19]:
# Personality type column is changed to one hot encoding
persona_type = pd.get_dummies(persona_type, columns=['Persona type'])
persona_type = persona_type.astype(int)
persona_type

Persona type_0  Persona type_1  Persona type_2  Persona type_3   
0                 0               0               1               0  \
1                 0               1               0               0   
2                 0               0               0               1   
3                 0               0               1               0   
4                 0               0               0               0   
..              ...             ...             ...             ...   
96                0               0               1               0   
97                0               0               1               0   
98                1               0               0               0   
99                1               0               0               0   
100               0               0               1               0   

     Persona type_4  
0                 0  
1                 0  
2                 0  
3                 0  
4                 1  
..              ...  
96                0  
97                0  
98                0  
99                0  
100               0  

[101 rows x 5 columns]

In [20]:
# Concatenate 'match_df' with 'persona_type'
match_df = pd.concat([match_df.iloc[:, :-25], persona_type], axis=1)
match_df

Genders  17-25  26-34  35-43  44-52  52+  Historical tours   
0          1      1      0      0      0    0                 0  \
1          1      1      0      0      0    0                 1   
2          1      1      0      0      0    0                 0   
3          0      0      0      0      0    0                 0   
4          0      1      0      0      0    0                 0   
..       ...    ...    ...    ...    ...  ...               ...   
96         1      1      0      0      0    0                 0   
97         1      1      0      0      0    0                 0   
98         1      1      0      0      0    0                 1   
99         1      0      1      0      0    0                 0   
100        0      1      0      0      0    0                 0   

     Adventure tours  Nature and wildlife tours  Culinary tours   
0                  1                          1               0  \
1                  0                          0               1   
2                  0                          1               1   
3                  1                          1               0   
4                  1                          1               1   
..               ...                        ...             ...   
96                 0                          1               1   
97                 1                          1               1   
98                 1                          0               1   
99                 1                          1               1   
100                1                          0               0   

     Wellness and retreat tours  Architectural tours  Educational tours   
0                             0                    0                  0  \
1                             0                    1                  0   
2                             0                    1                  0   
3                             0                    1                  0   
4                             0                    0                  0   
..                          ...                  ...                ...   
96                            0                    0                  0   
97                            0                    0                  0   
98                            0                    1                  0   
99                            0                    0                  0   
100                           0                    1                  0   

     Shopping tours  Persona type_0  Persona type_1  Persona type_2   
0                 1               0               0               1  \
1                 0               0               1               0   
2                 0               0               0               0   
3                 0               0               0               1   
4                 0               0               0               0   
..              ...             ...             ...             ...   
96                1               0               0               1   
97                0               0               0               1   
98                0               1               0               0   
99                0               1               0               0   
100               1               0               0               1   

     Persona type_3  Persona type_4  
0                 0               0  
1                 0               0  
2                 1               0  
3                 0               0  
4                 0               1  
..              ...             ...  
96                0               0  
97                0               0  
98                0               0  
99                0               0  
100               0               0  

[101 rows x 19 columns]

In [21]:
# Concatenate 'match_df' with 'persona_type'
user_df = pd.concat([user_df, persona_type], axis=1)
user_df

Names  Genders  Destinations  17-25  26-34  35-43   
0     Novia Rizki Wulandari         1             0      1      0      0  \
1    Herlina Kusyanuri Putri        1             0      1      0      0   
2               Rifdah Alyaa        1             0      1      0      0   
3             Alexis Purnomo        0             0      0      0      0   
4       Rifky Surya Pratama         0             0      1      0      0   
..                       ...      ...           ...    ...    ...    ...   
96                      Dera        1             0      1      0      0   
97          Salsabilla Rizki        1             0      1      0      0   
98                     Salma        1             0      1      0      0   
99                      Deva        1             0      0      1      0   
100                    Dzaky        0             0      1      0      0   

     44-52  52+  Historical tours  Adventure tours  ...  OPN2  OPN3  OPN4   
0        0    0                 0                1  ...     4     2     4  \
1        0    0                 1                0  ...     3     3     4   
2        0    0                 0                0  ...     2     3     4   
3        0    0                 0                1  ...     4     1     4   
4        0    0                 0                1  ...     3     3     2   
..     ...  ...               ...              ...  ...   ...   ...   ...   
96       0    0                 0                0  ...     5     5     5   
97       0    0                 0                1  ...     5     2     3   
98       0    0                 1                1  ...     5     4     5   
99       0    0                 0                1  ...     4     2     4   
100      0    0                 0                1  ...     5     5     4   

     OPN5  Roles  Persona type_0  Persona type_1  Persona type_2   
0       4  guide               0               0               1  \
1       3  guide               0               1               0   
2       3  guide               0               0               0   
3       4  guide               0               0               1   
4       3  guide               0               0               0   
..    ...    ...             ...             ...             ...   
96      5  guide               0               0               1   
97      4  guide               0               0               1   
98      4  guide               1               0               0   
99      3  guide               1               0               0   
100     5  guide               0               0               1   

     Persona type_3  Persona type_4  
0                 0               0  
1                 0               0  
2                 1               0  
3                 0               0  
4                 0               1  
..              ...             ...  
96                0               0  
97                0               0  
98                0               0  
99                0               0  
100               0               0  

[101 rows x 47 columns]

In [22]:
# Looking at the index of the columns
columns = user_df.columns
for idx, column in enumerate(columns):
    print(f"Column index: {idx}, Column name: {column}")

Column index: 0, Column name: Names
Column index: 1, Column name: Genders
Column index: 2, Column name: Destinations
Column index: 3, Column name: 17-25
Column index: 4, Column name: 26-34
Column index: 5, Column name: 35-43
Column index: 6, Column name: 44-52
Column index: 7, Column name: 52+
Column index: 8, Column name: Historical tours
Column index: 9, Column name: Adventure tours
Column index: 10, Column name: Nature and wildlife tours
Column index: 11, Column name: Culinary tours
Column index: 12, Column name: Wellness and retreat tours
Column index: 13, Column name: Architectural tours
Column index: 14, Column name: Educational tours
Column index: 15, Column name: Shopping tours
Column index: 16, Column name: EXT1
Column index: 17, Column name: EXT2
Column index: 18, Column name: EXT3
Column index: 19, Column name: EXT4
Column index: 20, Column name: EXT5
Column index: 21, Column name: EST1
Column index: 22, Column name: EST2
Column index: 23, Column name: EST3
Column index: 24,

# **Make Matchmaking Model**

In [23]:
match_array = match_df.values
match_array.shape

(101, 19)

## Build Function

In [24]:
def predict_pca(match_array):
    # Load the PCA model
    loaded_model = tf.saved_model.load('pca_model/1/')
    new_data_tf = tf.convert_to_tensor(match_array, dtype=tf.float32)

    # Predict the PCA value using the loaded model
    pca_value = loaded_model.apply_pca(new_data_tf)
    pca_value_arr = pca_value.numpy()
    
    return pca_value_arr

In [41]:
def get_guides_idx_filtered(user_id, user_df):
    # Get user destination and persona
    user_destination = user_df.loc[user_id, 'Destinations']
    user_persona = user_df.iloc[user_id, 42:].tolist()

    # Filter users with the same destination and roles value of 'guide'
    filtered_users = user_df[(user_df['Destinations'] == user_destination) & (user_df['Roles'] == 'guide')]

    # Filter based on personality
    user_indices = []
    for idx in filtered_users.index:
        if user_df.iloc[idx, 42:].tolist() == user_persona:
            user_indices.append(idx)
            
    # user_indices = pd.Index(filtered_users.index).tolist()
    return user_indices

In [37]:
def matchmaking(user_id, match_array, user_df):
    
    filtered_user_indices = get_guides_idx_filtered(user_id, user_df)
    if len(filtered_user_indices) == 0:
        print("No Results")
        sys.exit()
        
    user_indices = filtered_user_indices + [user_id]
    
    pca_data = predict_pca(match_array[user_indices])

    distances = []
    for idx in user_indices:
        # Get the distances within the cluster of the specified user

        distance_idx = euclidean_distances([pca_data[user_indices.index(user_id)]], [pca_data[user_indices.index(idx)]])
        distances.append(distance_idx)

    # Normalize the distances within the range of 0-100
    normalized_distances = 1 - distances / np.max(distances)
    scores = normalized_distances * 100
    
    # Reshape the list
    scores = np.reshape(scores, (len(user_indices)))
    
    matches = []
    for i in range(len(filtered_user_indices)):
        matches.append((filtered_user_indices[i], scores[i]))
    
    # Sort the matches based on the highest score
    matches.sort(key=lambda x: x[1], reverse=True)

    print("User Index:", user_id)
    print("Matched Peers:")
    match_idx = [user_id]
    for match_index, score in matches:
        if user_df.iloc[match_index, 41] == 'guide':
            if user_df.iloc[match_index, 2] == user_df.iloc[user_id, 2]:
                match_idx.append(match_index)
                print("Index:", match_index, "| Score:", f"{score:.2f}%")
                
    # Check if Destination and roles is correct 
    print()         
    print(user_df.iloc[match_idx, [0, 1, 2, 41]])
    
    # Check if they get same preference
    print()
    print(user_df.iloc[match_idx, 8:16])
    
    # Check the personality
    print()
    print(user_df.iloc[match_idx, 42:])

# **Test on New User**

## Build Function for New User

In [38]:
def predict_persona(new_user_persona_test):
    # Convert input data to float32
    data_input = np.array(new_user_persona_test, dtype=np.float32)

    # Load model
    model_path = 'persona_model/1'
    loaded_model = tf.saved_model.load(model_path)

    # Get the default signature of the model
    inference_fn = loaded_model.signatures["serving_default"]

    # Use models to make persona type predictions
    predictions = inference_fn(tf.constant(data_input))["dense_2"]
    
    new_user_persona_type = np.argmax(predictions, axis=1)
    new_user_persona_type = new_user_persona_type.tolist()

    binary_list = [[int(i == val) for i in range(5)] for val in new_user_persona_type]
    output_list = binary_list[0]
    return output_list 

In [39]:
def add_new_user(new_user, user_df):
    # Pick the result of personality test
    new_user_persona_test = new_user[16:41]

    # Convert to dataframe 
    new_user_persona_test = pd.DataFrame([new_user_persona_test ])
    
    # Menyisipkan result_persona setelah column roles
    new_user = new_user[:-1] + new_user[-1:] + predict_persona(new_user_persona_test)
    
    # Add the new row using loc indexer
    user_df.loc[len(user_df)] = new_user
    
    # Concatenate 'match_df' with 'persona_type'
    match_df = pd.concat([user_df.iloc[:, 1:2], user_df.iloc[:, 3:16], user_df.iloc[:, 42:]], axis=1)
    
    match_array = match_df.values
    
    return match_array

## Input New Data
We will input another user repeatly

In [54]:
new_user = ['Thariq', 1, 0, 1,0,0,0,0, 0,0,1,1,1,0,0,0, 4,5,1,2,3,5,3,4,4,4,1,1,2,4,3,2,4,1,2,3,2,5,5,1,1, 'tourist']

In [55]:
len(new_user)

42

In [56]:
match_array = add_new_user(new_user, user_df)

*before going to matchmaking, you can input again new user if u want*

In [57]:
user_df

Names  Genders  Destinations  17-25  26-34  35-43   
0     Novia Rizki Wulandari         1             0      1      0      0  \
1    Herlina Kusyanuri Putri        1             0      1      0      0   
2               Rifdah Alyaa        1             0      1      0      0   
3             Alexis Purnomo        0             0      0      0      0   
4       Rifky Surya Pratama         0             0      1      0      0   
..                       ...      ...           ...    ...    ...    ...   
100                    Dzaky        0             0      1      0      0   
101                   Thariq        1             0      1      0      0   
102                     Ravi        1             1      1      0      0   
103                   Fauzan        1             1      1      0      0   
104                   Thariq        1             0      1      0      0   

     44-52  52+  Historical tours  Adventure tours  ...  OPN2  OPN3  OPN4   
0        0    0                 0                1  ...     4     2     4  \
1        0    0                 1                0  ...     3     3     4   
2        0    0                 0                0  ...     2     3     4   
3        0    0                 0                1  ...     4     1     4   
4        0    0                 0                1  ...     3     3     2   
..     ...  ...               ...              ...  ...   ...   ...   ...   
100      0    0                 0                1  ...     5     5     4   
101      0    0                 0                0  ...     1     1     2   
102      0    0                 0                0  ...     1     1     4   
103      0    0                 0                0  ...     5     5     1   
104      0    0                 0                0  ...     5     5     1   

     OPN5    Roles  Persona type_0  Persona type_1  Persona type_2   
0       4    guide               0               0               1  \
1       3    guide               0               1               0   
2       3    guide               0               0               0   
3       4    guide               0               0               1   
4       3    guide               0               0               0   
..    ...      ...             ...             ...             ...   
100     5    guide               0               0               1   
101     3  tourist               1               0               0   
102     3  tourist               0               0               0   
103     1  tourist               0               0               0   
104     1  tourist               0               0               0   

     Persona type_3  Persona type_4  
0                 0               0  
1                 0               0  
2                 1               0  
3                 0               0  
4                 0               1  
..              ...             ...  
100               0               0  
101               0               0  
102               1               0  
103               1               0  
104               1               0  

[105 rows x 47 columns]

*If we see I add 4 new user, before we have **101 users** and now we have **105 users** so we will test all new users*

## Test Matching using New Users

In [43]:
matchmaking(101, match_array, user_df)

User Index: 101
Matched Peers:
Index: 30 | Score: 76.33%
Index: 14 | Score: 68.78%
Index: 29 | Score: 57.91%
Index: 91 | Score: 54.38%
Index: 55 | Score: 39.59%
Index: 45 | Score: 30.67%
Index: 99 | Score: 28.04%
Index: 51 | Score: 24.72%
Index: 46 | Score: 17.18%
Index: 98 | Score: 15.55%
Index: 13 | Score: 13.61%
Index: 42 | Score: 0.00%
Index: 85 | Score: 0.00%

                         Names  Genders  Destinations    Roles
101                     Thariq        1             0  tourist
30                     Farizka        1             0    guide
14   Dyah Ayu Kusuma Wardhani         1             0    guide
29                    Fernanda        1             0    guide
91                zahra kamila        1             0    guide
55                         Ney        1             0    guide
45                      Fauzan        0             0    guide
99                        Deva        1             0    guide
51                       Hafid        0             0    guide
46

In [48]:
matchmaking(102, match_array, user_df)

User Index: 102
Matched Peers:
Index: 86 | Score: 93.86%
Index: 36 | Score: 76.88%
Index: 81 | Score: 54.59%
Index: 26 | Score: 10.51%
Index: 11 | Score: 0.00%

                      Names  Genders  Destinations    Roles
102                    Ravi        1             1  tourist
86                  fadilah        1             1    guide
36            Fitrah Amalia        1             1    guide
81      Ihlasul Mufti Faqih        0             1    guide
26                    Nalen        0             1    guide
11   Muhammad Zidan Maulana        0             1    guide

     Historical tours  Adventure tours  Nature and wildlife tours   
102                 0                0                          1  \
86                  0                1                          0   
36                  0                0                          1   
81                  0                1                          1   
26                  0                1                          1   
11  

In [53]:
matchmaking(103, match_array, user_df)

User Index: 103
Matched Peers:
Index: 86 | Score: 93.86%
Index: 36 | Score: 76.88%
Index: 81 | Score: 54.59%
Index: 26 | Score: 10.51%
Index: 11 | Score: 0.00%

                      Names  Genders  Destinations    Roles
103                  Fauzan        1             1  tourist
86                  fadilah        1             1    guide
36            Fitrah Amalia        1             1    guide
81      Ihlasul Mufti Faqih        0             1    guide
26                    Nalen        0             1    guide
11   Muhammad Zidan Maulana        0             1    guide

     Historical tours  Adventure tours  Nature and wildlife tours   
103                 0                0                          1  \
86                  0                1                          0   
36                  0                0                          1   
81                  0                1                          1   
26                  0                1                          1   
11  

In [58]:
matchmaking(104, match_array, user_df)

User Index: 104
Matched Peers:
Index: 75 | Score: 82.99%
Index: 83 | Score: 75.03%
Index: 18 | Score: 71.24%
Index: 2 | Score: 70.56%
Index: 73 | Score: 57.08%
Index: 41 | Score: 36.08%
Index: 49 | Score: 30.20%
Index: 44 | Score: 29.91%
Index: 6 | Score: 28.52%
Index: 38 | Score: 26.68%
Index: 50 | Score: 18.08%
Index: 22 | Score: 6.16%
Index: 43 | Score: 4.63%
Index: 92 | Score: 0.00%

               Names  Genders  Destinations    Roles
104           Thariq        1             0  tourist
75           Maulani        1             0    guide
83            Elma N        1             0    guide
18               NCM        1             0    guide
2       Rifdah Alyaa        1             0    guide
73    Nabhan Nabilah        1             0    guide
41   Alyzar Aviandi         0             0    guide
49             Nadia        1             0    guide
44              Elsa        1             0    guide
6              fayza        1             0    guide
38     Daffa Baihaqi      